### PR TITLE
Visualisation Protocol Design and Specification

### LINKED ISSUE #443: Design Engine Visualisation Support
### Summary
The visualisations are the main USP of Enso Studio. In a world where every piece of data can be visualised seamlessly from inside the IDE, we need to ensure that we get support for this right.

To this end, we need to work out exactly how we plan to transport visualisation data between the runtime, engine and IDE, as well how the protocol messages for handling this are going to work.

### Value
Visualisations are an incredibly important selling point of Enso Studio, so quite obviously these have to work, and work well. 

### Specification

- [x] Determine exactly how the second transport for visualisation data is going to work. The suggestion and current design is for a binary web-socket. If this is kept, we need to determine how to point to segments of data sent across that socket.
- [x] Determine what protocol extensions we need to create to support visualisations:
  + The ability to create and destroy visualisation subscriptions with a given piece of Enso code as the preprocessing function.
  + The ability to update an existing subscription with a _new_ preprocessing function.
  + Visualisations should work on a pub/sub model, where an update is sent every time the underlying data is recomputed.
- [x] Specify these messages and the additional transport rigorously in the design document.

### Acceptance Criteria & Test Cases

- The above functionality has been implemented according to specification.
- The functionality has been rigorously tested.

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
This PR introduces the specification of the visualisation support messages in the engine protocol. In addition, it specifies the nature of the binary pipe's transport. 

It also adds additional fields to messages to provide additional detail to those implementing the protocol. 

Closes #443.

### Important Notes
This specification aims to be final, but it may change in minor ways as the feature is developed.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: doc/design/engine/engine-services.md [hunk 1]
```diff
@@ -580,6 +575,115 @@ and will be expanded upon as necessary in the future.
...
+used for transferring large amounts of data between Enso components.
+
+As the protocol is a binary transport, it is _mediated and controlled_ by
+messages that exist as part of the textual protocol.
+
+### Binary Protocol Communication Patterns
+The binary protocol currently only supports a single type of communication
+pattern:
+
+- **Push:** Messages containing data are pushed from the engine to the client in
```
Comment (kustosz) [inline doc/design/engine/engine-services.md:591]: The init message goes from client to server tho.

File: doc/design/engine/engine-services.md [hunk 2]
```diff
@@ -580,6 +575,115 @@ and will be expanded upon as necessary in the future.
...
+    [`session/initDataConnection`](#sessioninitdataconnection) below for more
+    information.
+
+## Service Connection Teardown
+As the engine performs sophisticated caching and persisting of data where
+possible, it is very important that the client informs the engine of the end of
+its session. In contrast to the initialisation flow above, this is not an
+involved process.
+
+1.  **Notify the Engine:** _Prior_ to disconnecting from the sockets, the client
```
Comment (kustosz) [inline doc/design/engine/engine-services.md:658]: Does it? Why do we care?

File: doc/design/engine/engine-services.md [hunk 3]
```diff
@@ -1057,13 +1173,102 @@ interface File {
...
+- **Direction:** Client -> Server
+- **Connection:** Data
+- **Visibility:** Public
+
+##### Parameters
+
+```idl
+namespace session;
+
+struct UUID {
```
Comment (kustosz) [inline doc/design/engine/engine-services.md:1231]: UUID is a 128-bit entity.

File: doc/design/engine/engine-services.md [hunk 4]
```diff
@@ -1234,6 +1449,45 @@ a given execution context.
...
+pre-serialized before being inserted into this message. As far as this level of
+transport is concerned, it is just a binary blob.
+
+##### Parameters
+
+```idl
+namespace executionContext;
+
+struct UUID {
+  id:uint64;
```
Comment (kustosz) [inline doc/design/engine/engine-services.md:1471]: Again, UUIDs are 128 bytes long.

File: doc/design/engine/engine-services.md (hunk 2 above)
Comment (iamrecursion) [reply doc/design/engine/engine-services.md:658]: So we know when all clients are disconnected and can save caches and things before shutting down.

File: doc/design/engine/engine-services.md [hunk 5]
```diff
@@ -1468,7 +1469,8 @@ transport is concerned, it is just a binary blob.
 namespace executionContext;
 
 struct UUID {
-  id:uint64;
+  firstHalf:uint64;
```
Comment (kustosz) [inline doc/design/engine/engine-services.md:1472]: maybe `highBytes` & `lowBytes` sounds better? Gives an endiannes-free interpretation.

**Pull request merged**